app/models/mixins/supports_feature_mixin.rb

```diff
@@ -132,7 +132,9 @@ def provider_classes_supporting(feature)
 
     # scope to query all those classes that support a particular feature
     def supporting(feature)
-      where(:type => subclasses_supporting(feature).map(&:name))
+      # First find all instances where the class supports <feature> then select instances
+      # which also support <feature> (e.g. the supports block does not add an unsupported_reason)
+      where(:type => subclasses_supporting(feature).map(&:name)).select { |instance| instance.supports?(feature) }
```

Comment:
I'm fine merging this as is, as it's really a refactoring of existing code, but I do share @kbrock's concern about performance.  As shown in your example, if all 14000 Vms would ultimately not be returned, it's fairly expensive to check them one-by-one, especially if the underlying supports method does an internal query (like checking a relationship or something).  That being said, maybe we just ensure the caller queries in the most efficient way?  For example, taking the example:

```ruby
>> Vm.supporting(:start).count
# 14336
```

The caller knows ahead of time (not programmatically, but just because they "know"), that powered on VMs can't be started, so they could rearrange their call to be

```
>> Vm.where.not(:power_state => "on").supporting(:start).count
=> 0
```

which essentially "pre-filters" them.

I wonder if there is a way to harness the knowledge in the supports call directly...perhaps similarly to how we do virtual attributes, or perhaps using virtual attributes directly?  Something like power_state easily becomes an arel check, which if called through virtual attributes, would automatically filter those out with a simple where clause.

Thinking something like the following (assuming a `supports_feature` virtual attribute exists)

```suggestion
      where(:type => subclasses_supporting(feature).map(&:name)).where(:supports_feature => true)
```

Reply:
Oh 100% agree instance level checks in supports? are terrible for performance, but that is how it has been designed since the beginning and we still need to use it that way unfortunately.  This doesn't change supports_feature is used currently since we're already doing instance level checks for things like buttons on the UI.

I don't know about pulling the logic out to pre-filter because that kind of defeats the purpose (the caller not needing to know what is needed, just asks if it is supported).

Definitely needs a redesign, have to keep in mind that different subclasses could have different requirements for instance-level checks.  Yeah was thinking instead of a block that has to be called it could return a scope, arel, or conditions for where or something like that.

Reply:
Ok yeah...this is sounding a lot like virtual attributes, which we should just use, especially since we are already exposing them over the API.